File: apps/stellacodeX/apple/StellaCodeX/StellaCodeX/Core/ServerProfile.swift
import Foundation

enum ServerConnectionMode: String, CaseIterable, Identifiable, Hashable, Codable {
    case direct
    case sshProxy = "ssh_proxy"

    var id: String {
        rawValue
    }

    var displayName: String {
        switch self {
        case .direct:
            "Direct"
        case .sshProxy:
            "SSH Proxy"
        }
    }
}

struct SSHProxyConfig: Hashable, Codable {
    var sshHost: String
    var sshPort: Int?
    var sshUser: String
    var targetURL: URL

    init(
        sshHost: String,
        sshPort: Int? = nil,
        sshUser: String = "",
        targetURL: URL
    ) {
        self.sshHost = sshHost
        self.sshPort = sshPort
        self.sshUser = sshUser
        self.targetURL = targetURL
    }

    var resolvedSSHPort: Int {
        sshPort ?? 22
    }
}

struct ServerProfile: Identifiable, Hashable, Codable {
    let id: UUID
    var name: String
    var connectionMode: ServerConnectionMode
    var baseURL: URL
    var sshProxy: SSHProxyConfig?
    var token: String
    var username: String

    init(
        id: UUID = UUID(),
        name: String,
        connectionMode: ServerConnectionMode = .direct,
        baseURL: URL,
        sshProxy: SSHProxyConfig? = nil,
        token: String = "",
        username: String
    ) {
        self.id = id
        self.name = name
        self.connectionMode = connectionMode
        self.baseURL = baseURL
        self.sshProxy = sshProxy
        self.token = token
        self.username = username
    }

    var targetURL: URL {
        sshProxy?.targetURL ?? baseURL
    }

    var connectionSummary: String {
        switch connectionMode {
        case .direct:
            return baseURL.absoluteString
        case .sshProxy:
            let host = sshProxy?.sshHost.trimmingCharacters(in: .whitespacesAndNewlines)
            let displayHost = host?.isEmpty == false ? host! : "missing SSH host"
            let user = sshProxy?.sshUser.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            let port = sshProxy?.sshPort.map { ":\($0)" } ?? ""
            let prefix = user.isEmpty ? "\(displayHost)\(port)" : "\(user)@\(displayHost)\(port)"
            return "\(prefix) -> \(targetURL.absoluteString)"
        }
    }
}

File: apps/stellacodeX/windows/src/StellaCodeX.Windows/Domain/Model/ServerProfile.cs
namespace StellaCodeX.Windows.Domain.Model;

public sealed record ServerProfile(
    Guid Id,
    string Name,
    ConnectionMode ConnectionMode,
    Uri BaseUrl,
    Uri TargetUrl,
    string SshHost,
    int? SshPort,
    string SshUser,
    string Token,
    string UserName)
{
    public Uri EffectiveTargetUrl => TargetUrl.AbsoluteUri.Length > 0 ? TargetUrl : BaseUrl;

    public bool IsConfigured => !string.IsNullOrWhiteSpace(Token) && ConnectionMode switch
    {
        ConnectionMode.Direct => IsUsableHttpUrl(BaseUrl),
        ConnectionMode.SshProxy => IsUsableHttpUrl(EffectiveTargetUrl) && !string.IsNullOrWhiteSpace(SshHost),
        _ => false,
    };

    public string ConnectionSummary => ConnectionMode switch
    {
        ConnectionMode.SshProxy => BuildSshSummary(),
        _ => TrimTrailingSlash(BaseUrl),
    };

    public static ServerProfile LocalDefault() => new(
        Id: Guid.NewGuid(),
        Name: "Local Stellaclaw",
        ConnectionMode: ConnectionMode.Direct,
        BaseUrl: new Uri("http://127.0.0.1:3111"),
        TargetUrl: new Uri("http://127.0.0.1:3111"),
        SshHost: string.Empty,
        SshPort: null,
        SshUser: string.Empty,
        Token: "local-web-token",
        UserName: "workspace-user");

    private string BuildSshSummary()
    {
        var host = string.IsNullOrWhiteSpace(SshHost) ? "missing SSH host" : SshHost.Trim();
        var userPrefix = string.IsNullOrWhiteSpace(SshUser) ? string.Empty : $"{SshUser.Trim()}@";
        var portSuffix = SshPort is > 0 ? $":{SshPort}" : string.Empty;
        return $"{userPrefix}{host}{portSuffix} -> {TrimTrailingSlash(EffectiveTargetUrl)}";
    }

    private static bool IsUsableHttpUrl(Uri uri) => uri.Scheme is "http" or "https";

    private static string TrimTrailingSlash(Uri uri) => uri.AbsoluteUri.TrimEnd('/');
}
